### PR TITLE
Add evaluation support for using lm-eval harness

### PR DESCRIPTION
**Pull request by shaahji**
## Add evaluation support for using lm-eval harness

* Added a new Evaluator to use lm_eval to evaluate hf model
* Added a llama2 config to use the new evaluator

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [x] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

File: examples/llama2/llama2_lmeval.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+    "input_model": {
+        "type": "HfModel",
+        "generative": true,
+        "model_path": "meta-llama/Llama-2-7b-hf",
+        "load_kwargs": { "attn_implementation": "eager" },
+        "io_config": {
```
[the remark this answers is not in the document]
Comment (shaahji) [reply examples/llama2/llama2_lmeval.json:7]: No, we don't. I will remove that.

File: examples/llama2/llama2_lmeval.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        "evaluator": {
+            "type": "LMEvaluator",
+            "model_class": "hf",
+            "tasks": [ "hellaswag" ],
+            "batch_size": 16,
+            "limit": 0.01,
+            "max_gen_toks": 10
+        }
+    },
+    "auto_optimizer_config": { "opt_level": 0, "disable_auto_optimizer": true, "precision": "fp16" },
```
[the remark this answers is not in the document]
Comment (shaahji) [reply examples/llama2/llama2_lmeval.json:35]: That's not true. If I don't explicitly disable the auto optimizer, I get this error -

```
ValueError: Search strategy is None but pass OrtTransformersOptimization has search space
```

[the remark this answers is not in the document]
Comment (devang-ml) [reply examples/llama2/llama2_lmeval.json:24]: @trajepl

[the remark this answers is not in the document]
Comment (shaahji) [reply examples/llama2/llama2_lmeval.json:24]: Why is the check for enabling auto-opt dependent on whether the config has passes or not?
https://github.com/microsoft/Olive/blob/1ceb55d53d5aaf0a7f5d6634fe83025af2c6158b/olive/workflows/run/run.py#L169-L173

**Review by devang-ml — APPROVED** [review]

File: examples/llama2/llama2_lmeval.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (trajepl) [reply examples/llama2/llama2_lmeval.json:24]: https://github.com/microsoft/Olive/pull/1013

There was a asking:
1. if input_model is given only, auto-opt should suggest the passes to optimize the input model by default without pass_config search by default
2. if input_model, evaluator, search_strategy are given, auto-opt should suggest pass flows and search pass configs by default

[the remark this answers is not in the document]
Comment (trajepl) [reply examples/llama2/llama2_lmeval.json:24]: > Why is the check for enabling auto-opt dependent on whether the config has passes or not?
> 
> https://github.com/microsoft/Olive/blob/1ceb55d53d5aaf0a7f5d6634fe83025af2c6158b/olive/workflows/run/run.py#L169-L173

In previous discussion, if no passes are give, enable auto-opt. If there are passes, just run the given passes.

**Pull request merged**